cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -377,52 +349,51 @@ private void publishKillProgram(String stopPayload) {
+  private ProgramStateUpdates getProgramStateUpdates(String peerName) {
+    String topic = programStateTopicPrefix + peerName;
+    String lastMessageId = lastMessageIdForPeer.get(peerName);
+    List<Notification> notifications = new ArrayList<>();
     try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
-                                                                    programUpdateTopic,
+                                                                    topic,
                                                                     programUpdateFetchSize,
-                                                                    afterMessageId)) {
+                                                                    lastMessageIdForPeer.get(peerName))) {
       while (iterator.hasNext()) {
         Message message = iterator.next();
-        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
-        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
-          Map<String, String> properties = notification.getProperties();
-          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
-          try {
-            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
-            if (runRecord.getPeerName() != null) {
-              String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
-              LOG.debug("Notifying peer {} about program run {} in state {}",
-                       runRecord.getPeerName(), programRunId, programStatus);
-              peerToNotifications.computeIfAbsent(runRecord.getPeerName(), n -> new ArrayList<>()).add(notification);
-            }
-          } catch (NotFoundException | IOException e) {
-            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
-          }
-        }
+        notifications.add(message.decodePayload(r -> GSON.fromJson(r, Notification.class)));
         lastMessageId = message.getId();
       }
-    } catch (Exception e) {
-      LOG.error("Exception when fetching program updates. Will retry again during next poll", e);
+    } catch (TopicNotFoundException e) {
+      createTopicIfNeeded(new TopicId(NamespaceId.SYSTEM.getNamespace(), topic));
+    } catch (IOException e) {
+      LOG.error("Exception when fetching program updates from per-peer topic. Will retry again during next poll", e);
+    }
+    return new ProgramStateUpdates(notifications, lastMessageId);
+  }
+
+  private void createTopicIfNeeded(TopicId topicId) {
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+      LOG.debug("Created topic {}", topicId.getTopic());
+    } catch (TopicAlreadyExistsException ex) {
+      // no-op
+    } catch (IOException ex) {
+      LOG.error("Failed to create topic {}", topicId.getTopic(), ex);
```

Comment:
instead of logging and returning, it's better to throw the IOException. The caller is the one who should decide what to do in error cases.

Reply:
this obsolete as topic creation logic has been removed on the subscriber side - handled by publisher